pkg/pss/evaluate.go

```diff
@@ -0,0 +1,1104 @@
+	for _, restrictedField := range restrictedFields {
+		if strings.Contains(restrictedField.path, "ontainers[*]") {
+			return true
+		}
+	}
+	return false
+}
+
+func ExemptProfile(checks []PSSCheckResult, rule *kyvernov1.PodSecurity, pod *corev1.Pod) (bool, error) {
+	ctx := enginectx.NewContext()
+
+	// 1. Iterate over check.RestrictedFields
+	// 2. Check if it's a `container-level` or `pod-level` restrictedField
+	// - `container-level`: container has a disallowed check (container name in check.ForbiddenDetail) && exempted by an exclude rule ? continue : pod creation is forbbiden
+	// - `pod-level`: Exempted by an exclude rule ? good : pod creation is forbbiden
+	for _, check := range checks {
+		for _, restrictedField := range check.RestrictedFields {
+			// Is a container-level restrictedField
+			if strings.Contains(restrictedField.path, "ontainers[*]") {
+				allowed, err := checkContainerLevelFields(ctx, pod, check, rule.Exclude, restrictedField)
+				if err != nil {
+					return false, errors.Wrap(err, err.Error())
+				}
+				if !allowed {
+					return false, nil
+				}
+			} else {
+				// Is a pod-level restrictedField
+				if !strings.Contains(check.CheckResult.ForbiddenDetail, "pod") && containsContainerLevelControl(check.RestrictedFields) {
+					continue
+				}
+				allowed, err := checkPodLevelFields(ctx, pod, check, rule, restrictedField)
+				if err != nil {
+					return false, errors.Wrap(err, err.Error())
+				}
+				if !allowed {
+					return false, nil
+				}
+			}
+		}
+	}
+	return true, nil
+}
+
+// Check if the pod creation is allowed after exempting some PSS controls
+func EvaluatePod(rule *kyvernov1.PodSecurity, pod *corev1.Pod, level *api.LevelVersion) (bool, []PSSCheckResult, error) {
+	// var podWithMatchingContainers corev1.Pod
+
+	// 1. Evaluate containers that match images specified in exclude
+	podWithMatchingContainers := getPodWithMatchingContainers(rule.Exclude, pod)
+
+	pssChecks := EvaluatePSS(level, &podWithMatchingContainers)
+
+	pssChecks = removePSSChecks(pssChecks, rule)
+
+	// 2. Check if all PSSCheckResults are exempted by exclude values
+	// Yes ? Evaluate pod's other containers
+	// No ? Pod creation forbidden
+	allowed, err := ExemptProfile(pssChecks, rule, &podWithMatchingContainers)
+	if err != nil {
+		return false, pssChecks, err
+	}
+	// Good to have: remove checks that are exempted and return only forbidden ones
+	if !allowed {
+		return false, pssChecks, nil
+	}
+
+	// 3. Optional, only when ExemptProfile() returns true
+	podWithNotMatchingContainers := getPodWithNotMatchingContainers(rule.Exclude, pod, &podWithMatchingContainers)
+	pssChecks = EvaluatePSS(level, &podWithNotMatchingContainers)
+	if len(pssChecks) > 0 {
+		return false, pssChecks, nil
+	}
+	return true, pssChecks, nil
+}
+
+func allowedValues(resourceValue interface{}, exclude kyvernov1.PodSecurityStandard, controls []restrictedField) bool {
+	for _, control := range controls {
```

Comment:
Can we extract this to a function and name it by its purpose? It's not easy to understand why it does here.